src/dynamic/HOmod_generator.cpp

```diff
@@ -0,0 +1,184 @@
+#include <cadmium/concept/coupled_model_assert.hpp>
+#include <cadmium/modeling/dynamic_coupled.hpp>
+#include <cadmium/modeling/dynamic_atomic.hpp>
+#include <cadmium/engine/pdevs_dynamic_runner.hpp>
+#include <cadmium/logger/tuple_to_ostream.hpp>
+#include <cadmium/logger/common_loggers.hpp>
+
+using TIME = float;
+
+// Ports for coupled models, we use the same in every level
+struct coupled_in_port1 : public cadmium::in_port<int>{};
+struct coupled_in_port2 : public cadmium::in_port<int>{};
+struct coupled_out_port : public cadmium::out_port<int>{};
+cadmium::dynamic::modeling::Ports coupled_in_ports = {typeid(coupled_in_port1), typeid(coupled_in_port2)};
+cadmium::dynamic::modeling::Ports coupled_out_ports = {typeid(coupled_out_port)};
+
+//A configured version of the devstone atomic, we use same configuration in every atomic.
+template<typename T>
+struct configured_atomic_devstone : devstone_atomic<T>{
+    configured_atomic_devstone(){
+        devstone_atomic<T>::period = 1;
+        devstone_atomic<T>::external_cycles = 100;
+        devstone_atomic<T>::internal_cycles = 100;
+    }
+};
+
+using ModelMatrix = std::vector<std::vector<std::shared_ptr<cadmium::dynamic::modeling::model>>>;
+
+std::shared_ptr<cadmium::dynamic::modeling::coupled<TIME>> create_HOmod_model(uint width, uint depth) {
+    // Creates the HOmod model with the passed parameters
+    // Returns a shared_ptr to the TOP model
+
+    //Level 0 has always a single model
+    std::shared_ptr<cadmium::dynamic::modeling::model> devstone_atomic_L0_0 = cadmium::dynamic::translate::make_dynamic_atomic_model<configured_atomic_devstone, TIME>("devstone_atomic_L0_0");
+    std::unordered_map<int, ModelMatrix> atomics_by_level;
+    std::unordered_map<int, std::shared_ptr<cadmium::dynamic::modeling::model>> coupleds_by_level;
+
+    for (int level=1; level <= depth; level++) {
+
+        //atomics
+        ModelMatrix atomics_current_level;
+        if (level < depth) {
+            //Last level does not have atomics
+            for(int idx_column=0; idx_column < width-1; idx_column++) {
+                cadmium::dynamic::modeling::Models models_col;
+                for(int idx_row=0; idx_row < idx_column + 2; idx_row++) {
+                    std::string atomic_name = "devstone_atomic_L" + std::to_string(level) + "_" + std::to_string(idx_column) + "," + std::to_string(idx_row);
+                    models_col.push_back(cadmium::dynamic::translate::make_dynamic_atomic_model<configured_atomic_devstone, TIME>(atomic_name));
+                }
+                atomics_current_level.push_back(models_col);
+            }
+        }
+        atomics_by_level[level] = atomics_current_level;
+
+        //coupled
+        cadmium::dynamic::modeling::Models Lcoupled_submodels;
+        cadmium::dynamic::modeling::EICs Lcoupled_eics;
+        cadmium::dynamic::modeling::EOCs Lcoupled_eocs;
+        cadmium::dynamic::modeling::ICs Lcoupled_ics = {};
+
+        if (level == 1) {
```

Comment:
It seems creating the first level is the same in all devstones, and the struct for creating it also the same. Maybe you can extract it into a different header file so we don't fail to edit it in a single place.